### PR TITLE
Split InternalAsset into UncommittedAsset and CommittedAsset (Cache ASTs followup)

### PR DESCRIPTION
**Pull request by wbinnssmith**
This is a followup based on my experience implementing parts of #3695, and also integrates some feedback from that PR.

This splits `InternalAsset` into `UncommittedAsset` and `CommittedAsset`. The former implements a mutable interface for transformers, only holding into values in memory (with the exception of generation), and the latter implements an immutable interface that loads content from disk and caches it in memory.

Previously, `InternalAsset` implemented both, in some situations branching on the state. This also allows for simpler implementations in each case. Common code was extracted into `assetUtils.js` in the same directory.

This also cleans up `babel7.js` to safely use `getAST()` and moves it to use the `Async` methods as we don't require synchronous behavior.

Test Plan: `yarn test`

### CODE REVIEW

**Comment by parcel-benchmark** [comment]
## Benchmark Results
<details><summary>packages/benchmarks/three 🚨</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | **FAILED** | -0.00ms |
| Cached | **FAILED** | -0.00ms |

#### Cold Bundles

*No bundles found, this is probably a failed build...*
#### Cached Bundles

*No bundles found, this is probably a failed build...*
</p></details>

<details><summary>packages/benchmarks/kitchen-sink 🚨</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | **FAILED** | -0.00ms |
| Cached | **FAILED** | -0.00ms |

#### Cold Bundles

*No bundles found, this is probably a failed build...*
#### Cached Bundles

*No bundles found, this is probably a failed build...*
</p></details>

<details><summary>packages/benchmarks/react-hn 🚨</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | **FAILED** | -0.00ms |
| Cached | **FAILED** | -0.00ms |

#### Cold Bundles

*No bundles found, this is probably a failed build...*
#### Cached Bundles

*No bundles found, this is probably a failed build...*
</p></details>

<details><summary>packages/benchmarks/ak-editor 🚨</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | **FAILED** | -0.00ms |
| Cached | **FAILED** | -0.00ms |

#### Cold Bundles

*No bundles found, this is probably a failed build...*
#### Cached Bundles

*No bundles found, this is probably a failed build...*
</p></details>

**[Click here to view a detailed benchmark overview.](https://benchmark.parceljs.org/benchmark/260940514099986945)**

File: packages/transformers/babel/package.json [hunk 1]
```diff
@@ -24,6 +24,7 @@
     "@babel/plugin-transform-typescript": "^7.4.5",
     "@babel/preset-env": "^7.0.0",
     "@babel/traverse": "^7.0.0",
+    "@parcel/babel-ast-utils": "^2.0.0-alpha.3.1",
```
Comment (wbinnssmith) [inline packages/transformers/babel/package.json]: It looks like the `import/no-extraneous-dependencies` lint rule didn't catch this. It seems to only be missed with packages in the monorepo?

File: packages/core/core/src/assetUtils.js [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+  let mapBuffer = map?.toBuffer();
+  // Store the results in the cache so we can avoid generating again next time
+  await Promise.all([
+    asset.options.cache.setStream(
+      nullthrows(asset.value.contentKey),
+      blobToStream(code),
+    ),
+    mapBuffer == null
+      ? Promise.resolve()
+      : asset.options.cache.setBlob(nullthrows(asset.value.mapKey), mapBuffer),
```
Comment (mischnic) [inline packages/core/core/src/assetUtils.js]:
```suggestion
    mapBuffer != null && asset.options.cache.setBlob(nullthrows(asset.value.mapKey), mapBuffer),
```

File: packages/core/core/src/CommittedAsset.js [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+  getMapBuffer(): Promise<?Buffer> {
+    let mapKey = this.value.mapKey;
+    if (mapKey != null && this.mapBuffer == null) {
+      this.mapBuffer = (async () => {
+        try {
+          return await this.options.cache.getBlob(mapKey);
+        } catch (err) {
+          if (err.code === 'ENOENT' && this.value.astKey != null) {
+            return (await generateFromAST(this)).map?.toBuffer();
+            // return output.map?.toBuffer();
```
Comment (devongovett) [inline packages/core/core/src/CommittedAsset.js]: Remove?

File: packages/core/core/src/assetUtils.js [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+  return {
+    code:
+      code instanceof Readable
+        ? asset.options.cache.getStream(nullthrows(asset.value.contentKey))
+        : code,
+    map,
+  };
+}
+
+export async function getConfig(
```
Comment (devongovett) [inline packages/core/core/src/assetUtils.js]: We should only expose this in transformers IMO

File: packages/core/core/src/assetUtils.js [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+    sideEffects: options.sideEffects != null ? options.sideEffects : true,
+    uniqueKey: uniqueKey,
+    plugin: options.plugin,
+    configPath: options.configPath,
+  };
+}
+
+const generateResults: WeakMap<Asset, Promise<GenerateOutput>> = new WeakMap();
+
+export function generateFromAST(
```
Comment (devongovett) [inline packages/core/core/src/assetUtils.js]: Is this only needed in CommittedAsset? Should we make this and the below function methods?

Comment (wbinnssmith) [reply packages/core/core/src/assetUtils.js]: Right now we generate early in `Transformation.js` if we're not scope hoisting, which has an `UncommittedAsset`.

File: packages/core/core/src/assetUtils.js (hunk 4 above)
Comment (wbinnssmith) [reply packages/core/core/src/assetUtils.js]: `getPackage` is probably useful everywhere though.

Comment (wbinnssmith) [reply packages/core/core/src/assetUtils.js]: Validators receive an `Asset` (not a `MutableAsset` unlike transformers), and use `.getConfig()` 😕

**Pull request merged**